ouroboros-network/src/Ouroboros/Network/Subscription/Dns.hs

```diff
@@ -0,0 +1,265 @@
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+
+
+module Ouroboros.Network.Subscription.Dns
+    ( DnsSubscriptionTarget (..)
+    , Resolver (..)
+    , dnsSubscriptionWorker'
+    , dnsSubscriptionWorker
+    , dnsResolve
+    , resolutionDelay
+    ) where
+
+import           Control.Monad (unless)
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSay
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadThrow
+import           Control.Monad.Class.MonadTime
+import           Control.Monad.Class.MonadTimer
+import           Control.Tracer
+import qualified Data.IP as IP
+import qualified Network.DNS as DNS
+import qualified Network.Socket as Socket
+import           Text.Printf
+
+import           Ouroboros.Network.Subscription.Common
+import           Ouroboros.Network.Subscription.Subscriber
+import           Ouroboros.Network.Socket
+
+-- | Time to wait for an AAAA response after receiving an A response.
+resolutionDelay :: DiffTime
+resolutionDelay = 0.05 -- 50ms delay
+
+data DnsSubscriptionTarget = DnsSubscriptionTarget {
+      dstDomain :: !DNS.Domain
+    , dstPort   :: !Socket.PortNumber
+    , dstValency :: !Int
+    } deriving (Eq, Show)
+
+data Resolver m = Resolver {
+      lookupA    :: DNS.Domain -> m (Either DNS.DNSError [Socket.SockAddr])
+    , lookupAAAA :: DNS.Domain -> m (Either DNS.DNSError [Socket.SockAddr])
+    }
+
+dnsResolve :: forall m.
+     ( MonadAsync m
+     , MonadSay   m
+     , MonadSTM   m
+     , MonadTime  m
+     , MonadTimer m
+     , MonadThrow m
+     )
+    => Tracer m DnsTrace
+    -> Resolver m
+    -> DnsSubscriptionTarget
+    -> m (SubscriptionTarget m Socket.SockAddr)
```

Comment:
Nice way of separating the rather complex policy we use here from the consumer of the targets.